### PR TITLE
Allow arbitrary non-nullable types as key for maps

### LINKED ISSUE #54537: Support complex types as map key
Currently OLAP systems like Spark and Presto supports complex types(array/map/tuple) as map key, but CH do not. 

I'm glad to contribute to it if you guys agree with this proposal.

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Maps can now have `Float32`, `Float64`, `Array(T)`, `Map(K,V)` and `Tuple(T1, T2, ...)` as keys. Closes #54537.

### CODE REVIEW

**Comment by taiyang-li** [comment]
Let's see what will happen in CI.

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 00b1ac53bd93c775b48cd2c2554ae3192af5e037 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59318/00b1ac53bd93c775b48cd2c2554ae3192af5e037/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59318/00b1ac53bd93c775b48cd2c2554ae3192af5e037/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by taiyang-li** [comment]
Hope for you reviews!

**Comment by alexey-milovidov** [comment]
@taiyang-li, please merge with master.

File: src/DataTypes/DataTypeMap.cpp [hunk 1]
```diff
@@ -124,11 +124,10 @@ bool DataTypeMap::checkKeyType(DataTypePtr key_type)
         if (!isStringOrFixedString(*(low_cardinality_data_type.getDictionaryType())))
             return false;
     }
-    else if (!key_type->isValueRepresentedByInteger()
-             && !isStringOrFixedString(*key_type)
-             && !WhichDataType(key_type).isNothing()
-             && !WhichDataType(key_type).isIPv6()
-             && !WhichDataType(key_type).isUUID())
+    else if (
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeMap.cpp:127]: It looks rather arbitrary. Why don't allow all types?

Comment (taiyang-li) [reply src/DataTypes/DataTypeMap.cpp:127]: I'll try

**Comment by taiyang-li** [comment]
> Please don't forget to update https://clickhouse.com/docs/en/sql-reference/data-types/map

done

**Comment by taiyang-li** [comment]
Failed uts seems not related to current changes. @rschu1ze 
- https://s3.amazonaws.com/clickhouse-test-reports/59318/a830462af2a9b1dabc1dbb1ee971d5b6587b6533/stateless_tests__tsan__s3_storage__[1_5].html
- https://s3.amazonaws.com/clickhouse-test-reports/59318/a830462af2a9b1dabc1dbb1ee971d5b6587b6533/stateless_tests__tsan__s3_storage__[1_5].html

**Comment by taiyang-li** [comment]
@rschu1ze  could you review it again? Thanks !

**Review by rschu1ze — COMMENTED** [review]
Looks generally good to me.

I made a few fix-ups locally (docs, exception message text, etc) but I don't have permission to push to your repo:

```
~/repo/ch (complex_map_key %>) $ git push bigo-sg complex_map_key
remote: Permission to bigo-sg/ClickHouse.git denied to rschu1ze.
fatal: unable to access 'https://github.com/bigo-sg/ClickHouse.git/': The requested URL returned error: 403
```

Anyways, I can push them in a separate PR once this is merged.

The msan error looks related, it is triggered by 02169_map_functions.sql, kindly check it.

**Comment by taiyang-li** [comment]
> Looks generally good to me.
> 
> I made a few fix-ups locally (docs, exception message text, etc) but I don't have permission to push to your repo:
> 
> ```
> ~/repo/ch (complex_map_key %>) $ git push bigo-sg complex_map_key
> remote: Permission to bigo-sg/ClickHouse.git denied to rschu1ze.
> fatal: unable to access 'https://github.com/bigo-sg/ClickHouse.git/': The requested URL returned error: 403
> ```
> 
> Anyways, I can push them in a separate PR once this is merged.
> 
> The msan error looks related, it is triggered by 02169_map_functions.sql, kindly check it.

Sorry to reply late. I rebuilt clickhouse with memory sanitizer locally, but it can't be reproduced.

**Comment by rschu1ze** [comment]
Sorry for the delay, just came back from vacation.

This query crashes with this PR (can be locally using a debug build).

The same failure appears [here](https://s3.amazonaws.com/clickhouse-test-reports/59318/5f786647fbc2c54dab9cd9885dffd4bfa4eec83a/ast_fuzzer__ubsan_.html).)

```
SELECT tuple(mapSort(k -> k, map([toUInt128(4), 3, 1], materialize(toNullable(4)), [4], 5))), mapUpdate(map('k1', 1, toFixedString('k2', 2), 2), map(toFixedString('k1', 2), materialize(toLowCardinality(11)), 'k2', 22)) GROUP BY toUInt256(3), tuple(toNullable(1)), 1 WITH CUBE;
```

-->

```
2024.04.02 13:05:55.889308 [ 327659 ] {d12508c5-c6c2-41d3-be3f-82f742a88f15} <Fatal> : Logical error: 'Bad cast from type DB::ColumnNode to DB::IdentifierNode'.
2024.04.02 13:05:55.890245 [ 327528 ] {} <Trace> BaseDaemon: Received signal 6
2024.04.02 13:05:55.890926 [ 329632 ] {} <Fatal> BaseDaemon: ########## Short fault info ############
2024.04.02 13:05:55.891048 [ 329632 ] {} <Fatal> BaseDaemon: (version 24.4.1.1, build id: F057518A4FE64982A1867F03A09D33B6FD800F28, git hash: 5f786647fbc2c54dab9cd9885dffd4bfa4eec83a) (from thread 327659) Received signal 6
2024.04.02 13:05:55.891106 [ 329632 ] {} <Fatal> BaseDaemon: Signal description: Aborted
2024.04.02 13:05:55.891138 [ 329632 ] {} <Fatal> BaseDaemon:
2024.04.02 13:05:55.891212 [ 329632 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007fc2a14969fc 0x00007fc2a1442476 0x00007fc2a14287f3 0x0000000012f8c703 0x0000000012f8c875 0x0000000012f8cdb1 0x000000000a5c560a 0x000000000a5c7eac 0x000000001a0d963c 0x000000001a0d9539 0x000000001a0d90b9 0x000000001ad75935 0x000000001ad7293f 0x000000001ad59cca 0x000000001ad580a1 0x000000001ad62241 0x000000001ad548d9 0x000000001ad52914 0x000000001ad525d3 0x000000001ad43d46 0x000000001b1bdb2c 0x0
00000001b1bce98 0x000000001b1bf818 0x000000001b1bef4c 0x000000001b1beefd 0x000000001b1bee9d 0x000000001b1bee5d 0x000000001b1bee25 0x000000001b0daa7e 0x000000001b0d5ffd 0x000000001b0d568d 0x000000001b85d250 0x000000001b857d4a 0x000000001d11b7c0 0x000000001d1327e5 0x000000002172cd99 0x000000002172d5e8 0x0000000021904094 0x0000000021900e3a 0x00000000218ffb7e 0x00007fc2a1494ac3
2024.04.02 13:05:55.891265 [ 329632 ] {} <Fatal> BaseDaemon: ########################################
2024.04.02 13:05:55.891436 [ 329632 ] {} <Fatal> BaseDaemon: (version 24.4.1.1, build id: F057518A4FE64982A1867F03A09D33B6FD800F28, git hash: 5f786647fbc2c54dab9cd9885dffd4bfa4eec83a) (from thread 327659) (query_id: d12508c5-c6c2-41d3-be3f-82f742a88f15) (query: SELECT tuple(mapSort(k -> k, map([toUInt128(4), 3, 1], materialize(toNullable(4)), [4], 5))), mapUpdate(map('k1', 1, toFixedString('k2', 2), 2), map(toFixedString('k1', 2), materialize(toLowCardinality(11)), 'k2',
22)) GROUP BY toUInt256(3), tuple(toNullable(1)), 1 WITH CUBE) Received signal Aborted (6)
2024.04.02 13:05:55.891549 [ 329632 ] {} <Fatal> BaseDaemon:
2024.04.02 13:05:55.891652 [ 329632 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007fc2a14969fc 0x00007fc2a1442476 0x00007fc2a14287f3 0x0000000012f8c703 0x0000000012f8c875 0x0000000012f8cdb1 0x000000000a5c560a 0x000000000a5c7eac 0x000000001a0d963c 0x000000001a0d9539 0x000000001a0d90b9 0x000000001ad75935 0x000000001ad7293f 0x000000001ad59cca 0x000000001ad580a1 0x000000001ad62241 0x000000001ad548d9 0x000000001ad52914 0x000000001ad525d3 0x000000001ad43d46 0x000000001b1bdb2c 0x0
00000001b1bce98 0x000000001b1bf818 0x000000001b1bef4c 0x000000001b1beefd 0x000000001b1bee9d 0x000000001b1bee5d 0x000000001b1bee25 0x000000001b0daa7e 0x000000001b0d5ffd 0x000000001b0d568d 0x000000001b85d250 0x000000001b857d4a 0x000000001d11b7c0 0x000000001d1327e5 0x000000002172cd99 0x000000002172d5e8 0x0000000021904094 0x0000000021900e3a 0x00000000218ffb7e 0x00007fc2a1494ac3
2024.04.02 13:05:55.891775 [ 329632 ] {} <Fatal> BaseDaemon: 4. ? @ 0x00007fc2a14969fc
2024.04.02 13:05:55.891850 [ 329632 ] {} <Fatal> BaseDaemon: 5. ? @ 0x00007fc2a1442476
2024.04.02 13:05:55.891958 [ 329632 ] {} <Fatal> BaseDaemon: 6. ? @ 0x00007fc2a14287f3
2024.04.02 13:05:55.960419 [ 329632 ] {} <Fatal> BaseDaemon: 7. /home/ubuntu/repo/ch7/src/Common/Exception.cpp:0: DB::abortOnFailedAssertion(String const&) @ 0x0000000012f8c703
2024.04.02 13:05:56.005481 [ 328435 ] {} <Trace> AsynchronousMetrics: MemoryTracking: was 640.55 MiB, peak 116.97 GiB, free memory in arenas 43.50 MiB, will set to 669.56 MiB (RSS), difference: 29.01 MiB
2024.04.02 13:05:56.027512 [ 329632 ] {} <Fatal> BaseDaemon: 8. /home/ubuntu/repo/ch7/src/Common/Exception.cpp:63: DB::handle_error_code(String const&, int, bool, std::vector<void*, std::allocator<void*>> const&) @ 0x0000000012f8c875
2024.04.02 13:05:56.083235 [ 329632 ] {} <Fatal> BaseDaemon: 9. /home/ubuntu/repo/ch7/src/Common/Exception.cpp:100: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000012f8cdb1
2024.04.02 13:05:56.132762 [ 329632 ] {} <Fatal> BaseDaemon: 10. /home/ubuntu/repo/ch7/src/Common/Exception.h:90: DB::Exception::Exception(String&&, int, bool) @ 0x000000000a5c560a
2024.04.02 13:05:56.183431 [ 329632 ] {} <Fatal> BaseDaemon: 11. /home/ubuntu/repo/ch7/src/Common/Exception.h:109: DB::Exception::Exception<String, String>(int, FormatStringHelperImpl<std::type_identity<String>::type, std::type_identity<String>::type>, String&&, String&&) @ 0x000000000a5c7eac
2024.04.02 13:05:56.281722 [ 329632 ] {} <Fatal> BaseDaemon: 12. /home/ubuntu/repo/ch7/src/Common/typeid_cast.h:33: DB::IdentifierNode& typeid_cast<DB::IdentifierNode&, DB::IQueryTreeNode>(DB::IQueryTreeNode&) @ 0x000000001a0d963c
2024.04.02 13:05:56.378949 [ 329632 ] {} <Fatal> BaseDaemon: 13. /home/ubuntu/repo/ch7/src/Common/TypePromotion.h:26: DB::TypePromotion<DB::IQueryTreeNode>::CastHelper<DB::IdentifierNode&, false, true>::value(DB::IQueryTreeNode*) @ 0x000000001a0d9539
2024.04.02 13:05:56.475800 [ 329632 ] {} <Fatal> BaseDaemon: 14. /home/ubuntu/repo/ch7/src/Common/TypePromotion.h:52: std::invoke_result<decltype(&CastHelper<DB::IdentifierNode&, false>::value), DB::TypePromotion<DB::IQueryTreeNode>::CastHelper<DB::IdentifierNode&, false, std::is_reference_v<DB::IdentifierNode&>>, DB::IQueryTreeNode*>::type DB::TypePromotion<DB::IQueryTreeNode>::as<DB::IdentifierNode&>() @ 0x000000001a0d90b9
2024.04.02 13:05:56.876662 [ 329632 ] {} <Fatal> BaseDaemon: 15. /home/ubuntu/repo/ch7/src/Analyzer/Passes/QueryAnalysisPass.cpp:5185: DB::(anonymous namespace)::QueryAnalyzer::resolveLambda(std::shared_ptr<DB::IQueryTreeNode> const&, std::shared_ptr<DB::IQueryTreeNode> const&, std::vector<std::shared_ptr<DB::IQueryTreeNode>, std::allocator<std::shared_ptr<DB::IQueryTreeNode>>> const&, DB::(anonymous namespace)::IdentifierResolveScope&) @ 0x000000001ad75935
2024.04.02 13:05:57.005776 [ 328435 ] {} <Trace> AsynchronousMetrics: MemoryTracking: was 669.56 MiB, peak 116.97 GiB, free memory in arenas 43.48 MiB, will set to 764.56 MiB (RSS), difference: 95.00 MiB
2024.04.02 13:05:57.218551 [ 329632 ] {} <Fatal> BaseDaemon: 16. /home/ubuntu/repo/ch7/src/Analyzer/Passes/QueryAnalysisPass.cpp:5932: DB::(anonymous namespace)::QueryAnalyzer::resolveFunction(std::shared_ptr<DB::IQueryTreeNode>&, DB::(anonymous namespace)::IdentifierResolveScope&) @ 0x000000001ad7293f
2024.04.02 13:05:57.575496 [ 329632 ] {} <Fatal> BaseDaemon: 17. /home/ubuntu/repo/ch7/src/Analyzer/Passes/QueryAnalysisPass.cpp:6374: DB::(anonymous namespace)::QueryAnalyzer::resolveExpressionNode(std::shared_ptr<DB::IQueryTreeNode>&, DB::(anonymous namespace)::IdentifierResolveScope&, bool, bool) @ 0x000000001ad59cca
2024.04.02 13:05:57.937301 [ 329632 ] {} <Fatal> BaseDaemon: 18. /home/ubuntu/repo/ch7/src/Analyzer/Passes/QueryAnalysisPass.cpp:6509: DB::(anonymous namespace)::QueryAnalyzer::resolveExpressionNodeList(std::shared_ptr<DB::IQueryTreeNode>&, DB::(anonymous namespace)::IdentifierResolveScope&, bool, bool) @ 0x000000001ad580a1
2024.04.02 13:05:58.005783 [ 328435 ] {} <Trace> AsynchronousMetrics: MemoryTracking: was 764.56 MiB, peak 116.97 GiB, free memory in arenas 44.71 MiB, will set to 765.56 MiB (RSS), difference: 1.00 MiB
2024.04.02 13:05:58.306387 [ 329632 ] {} <Fatal> BaseDaemon: 19. /home/ubuntu/repo/ch7/src/Analyzer/Passes/QueryAnalysisPass.cpp:6720: DB::(anonymous namespace)::QueryAnalyzer::resolveGroupByNode(DB::QueryNode&, DB::(anonymous namespace)::IdentifierResolveScope&) @ 0x000000001ad62241
^C2024.04.02 13:05:58.325640 [ 327528 ] {} <Trace> BaseDaemon: Received signal 2
2024.04.02 13:05:58.325711 [ 327528 ] {} <Information> Application: Received termination signal (Interrupt)
2024.04.02 13:05:58.325884 [ 327527 ] {} <Debug> Application: Received termination signal.
2024.04.02 13:05:58.326609 [ 327527 ] {} <Debug> Application: Waiting for current connections to close.
2024.04.02 13:05:58.667199 [ 329632 ] {} <Fatal> BaseDaemon: 20. /home/ubuntu/repo/ch7/src/Analyzer/Passes/QueryAnalysisPass.cpp:8032: DB::(anonymous namespace)::QueryAnalyzer::resolveQuery(std::shared_ptr<DB::IQueryTreeNode> const&, DB::(anonymous namespace)::IdentifierResolveScope&) @ 0x000000001ad548d9
2024.04.02 13:05:59.005358 [ 328435 ] {} <Trace> AsynchronousMetrics: MemoryTracking: was 765.55 MiB, peak 116.97 GiB, free memory in arenas 42.21 MiB, will set to 763.71 MiB (RSS), difference: -1.83 MiB
2024.04.02 13:05:59.024083 [ 329632 ] {} <Fatal> BaseDaemon: 21. /home/ubuntu/repo/ch7/src/Analyzer/Passes/QueryAnalysisPass.cpp:1127: DB::(anonymous namespace)::QueryAnalyzer::resolve(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::IQueryTreeNode> const&, std::shared_ptr<DB::Context const>) @ 0x000000001ad52914
2024.04.02 13:05:59.380590 [ 329632 ] {} <Fatal> BaseDaemon: 22. /home/ubuntu/repo/ch7/src/Analyzer/Passes/QueryAnalysisPass.cpp:8227: DB::QueryAnalysisPass::run(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::Context const>) @ 0x000000001ad525d3
2024.04.02 13:05:59.389173 [ 327527 ] {} <Warning> Application: Closed all listening sockets. Waiting for 1 outstanding connections.
2024.04.02 13:05:59.450007 [ 329632 ] {} <Fatal> BaseDaemon: 23. /home/ubuntu/repo/ch7/src/Analyzer/QueryTreePassManager.cpp:186: DB::QueryTreePassManager::run(std::shared_ptr<DB::IQueryTreeNode>) @ 0x000000001ad43d46
2024.04.02 13:05:59.549388 [ 329632 ] {} <Fatal> BaseDaemon: 24. /home/ubuntu/repo/ch7/src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:115: DB::(anonymous namespace)::buildQueryTreeAndRunPasses(std::shared_ptr<DB::IAST> const&, DB::SelectQueryOptions const&, std::shared_ptr<DB::Context const> const&, std::shared_ptr<DB::IStorage> const&) @ 0x000000001b1bdb2c
2024.04.02 13:05:59.632909 [ 329632 ] {} <Fatal> BaseDaemon: 25. /home/ubuntu/repo/ch7/src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:132: DB::InterpreterSelectQueryAnalyzer::InterpreterSelectQueryAnalyzer(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const> const&, DB::SelectQueryOptions const&) @ 0x000000001b1bce98
```

The performance failures (e.g. [here](https://s3.amazonaws.com/clickhouse-test-reports/59318/5f786647fbc2c54dab9cd9885dffd4bfa4eec83a/performance_comparison_[1_4]/report.html)) are unrelated.

Do you also like to take a look here which looks related (on the surface)? https://s3.amazonaws.com/clickhouse-test-reports/59318/5f786647fbc2c54dab9cd9885dffd4bfa4eec83a/ast_fuzzer__tsan_.html

The test failure of 00062_loyalty ([here](https://s3.amazonaws.com/clickhouse-test-reports/59318/5f786647fbc2c54dab9cd9885dffd4bfa4eec83a/stateful_tests__asan_.html)) looks also weird but it does not contain `maps`, so I assume it is unrelated.

Please also merge once more from `master` as lots of fixes for CI were merged in the meantime.

**Comment by pufit** [comment]
> The same failure appears [here](https://s3.amazonaws.com/clickhouse-test-reports/59318/5f786647fbc2c54dab9cd9885dffd4bfa4eec83a/ast_fuzzer__ubsan_.html).)

BTW, there is the same logical error in the master. Probably, it's related to the new analyzer being merged
https://s3.amazonaws.com/clickhouse-test-reports/0/0fe3046a985fe9ab566422ef10fae20fc5c4d8c4/ast_fuzzer__ubsan_.html

**Comment by taiyang-li** [comment]
> Do you also like to take a look here which looks related (on the surface)? https://s3.amazonaws.com/clickhouse-test-reports/59318/5f786647fbc2c54dab9cd9885dffd4bfa4eec83a/ast_fuzzer__tsan_.html 

@rschu1ze  I tried, but it could not be reproduced.

**Comment by rschu1ze** [comment]
> > The same failure appears [here](https://s3.amazonaws.com/clickhouse-test-reports/59318/5f786647fbc2c54dab9cd9885dffd4bfa4eec83a/ast_fuzzer__ubsan_.html).)
> 
> BTW, there is the same logical error in the master. Probably, it's related to the new analyzer being merged https://s3.amazonaws.com/clickhouse-test-reports/0/0fe3046a985fe9ab566422ef10fae20fc5c4d8c4/ast_fuzzer__ubsan_.html

```sql
SELECT tuple(mapSort(k -> k, map([toUInt128(4), 3, 1], materialize(toNullable(4)), [4], 5))), mapUpdate(map('k1', 1, toFixedString('k2', 2), 2), map(toFixedString('k1', 2), materialize(toLowCardinality(11)), 'k2', 22)) GROUP BY toUInt256(3), tuple(toNullable(1)), 1 WITH CUBE;
```

does not run in `master` but a minimized version which triggers the same error does:

```sql
SELECT tuple(mapSort(k -> k, map('a', 4))), mapUpdate(map('k1', 1), map('k2', 11)) GROUP BY 1 WITH CUBE;
```

Opened issue #62374.

There are still many analyzer-related failures :-( Let's wait a bit more until the CI situation is better.

**Comment by rschu1ze** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/59318/c0ddd3df0bc37d0b842073fb247ef4f3b64d4a49/ast_fuzzer__ubsan_.html:
- `Logical error: 'Columns are assumed to be of identical types, but they are different in Nullable'.`.
- This was caused by query `SELECT 1, map((1, 2), materialize(4), (3, 4), 5), mapFilter((k, v) -> (k = (materialize(1), 2)), map((1, 2), toUInt256(4), (3, 4), 5)) GROUP BY tuple('b', 1, 1, materialize(1), toNullable(1), 1, toLowCardinality(1), toNullable(toNullable(1)), 1), CAST('0.1', 'Float32')`.
- Running this query on `master` and on this `PR` did not reproduce the error. I did not find registered occurrences of this error, so it is a bit strange but I assume it is a one-hit-wonder.

I'll merge and keep an eye on the error.

**Comment by den-crane** [comment]
* https://github.com/ClickHouse/ClickHouse/issues/64645

**Comment by clickgapai** [comment]
Hi @taiyang-li @rschu1ze — while reviewing this PR I found the following:

- **Bug (P1):** VALUES clause silently produces empty Map when inserting `Map(Float64, V)` into `Map(Float32, V)` column — https://github.com/ClickHouse/ClickHouse/issues/104438

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**